### PR TITLE
Do not publish exposed ports in Docker machines automatically

### LINKED ISSUE #6124: Ports should not be exposed automatically, but rather be explicitly defined in workspace config
Related to SPI.
Machine info from workspace config must contain information about the exposed port in `servers` property:

```
    "workspaceConfig": {
      "environments": {
        "default": {
          "machines": {
            "app/main": {
              "installers": [
                "org.eclipse.che.exec", "org.eclipse.che.terminal", "org.eclipse.che.ws-agent"
              ],
              "servers": {
                "web": {
                  "protocol": "http",
                  "port": "8080/tcp"
                }
              }
            }
.
.
.
```
This would make it easier and more straight forward creating workspaces on different infrastructures ( e.g. no need for inspecting image while creating workspace against `dockerimage` recipe)

### PR DESCRIPTION
**Pull request by garagatyi**
### What does this PR do?
Publish port in Docker machines only in case there is a server
in the workspace config that represents the port.
Refactor OpenShift infrastructure code.
Refactor Docker infrastructure code.
Add constant of memory attribute and reuse it everywhere it is
hardcoded.
Add constant for installer environment attribute and reuse it
everywhere it is hardcoded.
Add machine memory attribute validation to workspace validation.
Move provisioning of servers, environment variables from
infrastructure implementations to runtime SPI abstract level.

### What has to be done yet?
Migrate stacks/factories in DB.

### What issues does this PR fix or reference?
Fixes #6124 

#### Release Notes

Ports that are exposed in a Docker image/Dockerfile/Compose file are no longer opened to the external access automatically. To open a port, please, add Server into workspace definition.
This is a backward incompatible change that will disable access from a browser to some ports that were opened in Che 5.

### CODE REVIEW

File: infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/provision/labels/RuntimeLabelsProvisioner.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+import java.util.HashMap;
+import java.util.Map;
+import org.eclipse.che.api.core.model.workspace.runtime.RuntimeIdentity;
+import org.eclipse.che.api.workspace.server.spi.InfrastructureException;
+import org.eclipse.che.api.workspace.server.spi.InternalEnvironment;
+import org.eclipse.che.workspace.infrastructure.openshift.environment.OpenShiftEnvironment;
+import org.eclipse.che.workspace.infrastructure.openshift.provision.ConfigurationProvisioner;
+
+/** @author Alexander Garagatyi */
+public class RuntimeLabelsProvisioner implements ConfigurationProvisioner {
```
Comment (sleshchenko) [inline infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/provision/labels/RuntimeLabelsProvisioner.java]: I think it is not `RuntimeLabelsProvisioner`. Maybe `PodNameLabelProvisioner` would be better.

File: infrastructures/docker/infrastructure/src/main/java/org/eclipse/che/workspace/infrastructure/docker/provisioner/server/ServersConverter.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+import org.eclipse.che.api.core.model.workspace.runtime.RuntimeIdentity;
+import org.eclipse.che.api.workspace.server.spi.InfrastructureException;
+import org.eclipse.che.api.workspace.server.spi.InternalEnvironment;
+import org.eclipse.che.workspace.infrastructure.docker.InfrastructureProvisioner;
+import org.eclipse.che.workspace.infrastructure.docker.Labels;
+import org.eclipse.che.workspace.infrastructure.docker.model.DockerContainerConfig;
+import org.eclipse.che.workspace.infrastructure.docker.model.DockerEnvironment;
+
+/** @author Alexander Garagatyi */
```
Comment (sleshchenko) [inline infrastructures/docker/infrastructure/src/main/java/org/eclipse/che/workspace/infrastructure/docker/provisioner/server/ServersConverter.java]: Information about what converting means could be useful. Please consider adding java doc here about putting required labels/ports/exposes.

File: infrastructures/docker/infrastructure/src/main/java/org/eclipse/che/workspace/infrastructure/docker/local/LocalCheInfrastructureProvisioner.java [hunk 3]
```diff
@@ -34,48 +37,65 @@
...
-      LocalInstallersConfigProvisioner installerConfigProvisioner,
-      LabelsProvisioner labelsProvisioner,
+      LocalInstallersBinariesVolumeProvisioner installersBinariesVolumeProvisioner,
+      RuntimeLabelsProvisioner runtimeLabelsProvisioner,
       DockerApiHostEnvVariableProvisioner dockerApiEnvProvisioner,
-      ToolingServersEnvVarsProvisioner toolingServersEnvVarsProvisioner,
-      WsAgentServerConfigProvisioner wsAgentServerConfigProvisioner) {
+      ToolingServersEnvVarsProvisioner agentsServersEnvVarsProvisioner,
+      WsAgentServerConfigProvisioner wsAgentServerConfigProvisioner,
+      ServersConverter serversToPortsConverter,
```
Comment (sleshchenko) [inline infrastructures/docker/infrastructure/src/main/java/org/eclipse/che/workspace/infrastructure/docker/local/LocalCheInfrastructureProvisioner.java]: Please check variables names here

**Review by sleshchenko — APPROVED** [review]
I like your refactoring so much! I think it makes provisioning logic more clear :+1: 
One thing which can be useful for review or future git history usage - splitting changes related to exposing ports, renaming agent to installers in cli and refactoring in different commits

**Review by benoitf — APPROVED** [review]

File: infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/provision/labels/RuntimeLabelsProvisioner.java (hunk 1 above)
Comment (garagatyi) [reply infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/provision/labels/RuntimeLabelsProvisioner.java]: ok

File: infrastructures/docker/infrastructure/src/main/java/org/eclipse/che/workspace/infrastructure/docker/provisioner/server/ServersConverter.java (hunk 2 above)
Comment (garagatyi) [reply infrastructures/docker/infrastructure/src/main/java/org/eclipse/che/workspace/infrastructure/docker/provisioner/server/ServersConverter.java]: ok

File: infrastructures/docker/infrastructure/src/main/java/org/eclipse/che/workspace/infrastructure/docker/local/LocalCheInfrastructureProvisioner.java (hunk 3 above)
Comment (garagatyi) [reply infrastructures/docker/infrastructure/src/main/java/org/eclipse/che/workspace/infrastructure/docker/local/LocalCheInfrastructureProvisioner.java]: Good catch, tnx

**Comment by garagatyi** [comment]
ci-test

**Review by skabashnyuk — APPROVED** [review]

**Review by ashumilova — APPROVED** [review]

**Review by akorneta — APPROVED** [review]

**Review by skabashnyuk — APPROVED** [review]

**Review by riuvshin — APPROVED** [review]

**Comment by musienko-maxim** [comment]
The functional tests did not find regression

File: selenium/che-selenium-test/src/test/resources/templates/workspace/docker/ubuntu_with_c_sharp_lsp.json [hunk 4]
```diff
@@ -3,7 +3,20 @@
     "replaced_name": {
       "machines": {
         "dev-machine": {
-          "servers": {},
+          "servers": {
```
[the remark this answers is not in the document]
Comment (sleshchenko) [reply selenium/che-selenium-test/src/test/resources/templates/workspace/docker/ubuntu_with_c_sharp_lsp.json:6]: I mean updating of [ws-agent.json](https://github.com/eclipse/che/pull/6920/files#diff-3cdb9abacf2198ec648da0d48e371bf1R22)

[the remark this answers is not in the document]
Comment (garagatyi) [reply selenium/che-selenium-test/src/test/resources/templates/workspace/docker/ubuntu_with_c_sharp_lsp.json:6]: @sleshchenko @tolusha @gazarenkov Should I introduce wsagent installer v 1.0.1 in that case?

[the remark this answers is not in the document]
Comment (tolusha) [reply selenium/che-selenium-test/src/test/resources/templates/workspace/docker/ubuntu_with_c_sharp_lsp.json:6]: Yes, add a new ws-installer v1.0.1

[the remark this answers is not in the document]
Comment (garagatyi) [reply selenium/che-selenium-test/src/test/resources/templates/workspace/docker/ubuntu_with_c_sharp_lsp.json:6]: Added

**Pull request merged**